cis194/week5/tien/src/Calc.hs

```diff
@@ -0,0 +1,74 @@
+{-# LANGUAGE TypeSynonymInstances #-}
+{-# OPTIONS_GHC -Wall #-}
+{-# LANGUAGE ViewPatterns         #-}
+
+module Calc where
+
+import           ExprT
+import           Parser
+import           StackVM
+
+-- Exercise 1 --
+eval :: ExprT -> Integer
+eval (Lit a)         = a
+eval (ExprT.Add a b) = eval a + eval b
+eval (ExprT.Mul a b) = eval a * eval b
+
+-- Exercise 2 --
+evalStr :: String -> Maybe Integer
+evalStr (parseExp Lit ExprT.Add ExprT.Mul -> justExpr) = eval <$> justExpr
```

Comment:
We are merely binding the result to a name `justExpr`. This means we are not using any view patterns features.
```suggestion
evalStr = fmap eval . parseExp Lit ExprT.Add ExprT.Mul
```